packages/core/src/browser/breadcrumbs/breadcrumbs-service.ts

```diff
@@ -0,0 +1,73 @@
+/********************************************************************************
+ * Copyright (C) 2019 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable, named } from 'inversify';
+import { ContributionProvider, Prioritizeable, Emitter, Event } from '../../common';
+import URI from '../../common/uri';
+import { Breadcrumb } from './breadcrumb';
+import { BreadcrumbPopup } from './breadcrumb-popup';
+import { BreadcrumbsContribution } from './breadcrumbs-contribution';
+
+@injectable()
+export class BreadcrumbsService {
+
+    @inject(ContributionProvider) @named(BreadcrumbsContribution)
+    protected readonly contributions: ContributionProvider<BreadcrumbsContribution>;
+
+    protected readonly onBreadcrumbsChangeEmitter = new Emitter<URI>();
+
+    /**
+     * Subscribe to this event emitter to be notifed when the breadcrumbs have changed.
+     * The URI is the URI of the editor the breadcrumbs have changed for.
+     */
+    get onBreadcrumbsChange(): Event<URI> {
+        return this.onBreadcrumbsChangeEmitter.event;
+    }
+
+    /**
+     * Notifies that the breadcrumbs for the given URI have changed and should be re-rendered.
+     * This fires an `onBreadcrumsChange` event.
+     */
+    breadcrumbsChanges(uri: URI): void {
```

Comment:
firing events should be an implementation detail, i.e. `BreadcrumbsContribution` should have `onDidChange` instead

Reply:
Maybe I don't get it. But it sounds that this would make it more complicated without any substantial benefit since the breadcrumbs renderer has to subscribe to all breadcrumb contributions then.